# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/events/Event.cpp
# Commit: f1e7580708c9
# Full Hash: f1e7580708c941ff3f6ee395462864f1645ef74e
# Author: Olli Pettay <Olli.Pettay@helsinki.fi>
# Date: 2025-01-11 09:16:48
# Regressor Bug: 1926530
# File Overlap Count: 1
# Description:
#   Bug 1926530 - skip certainly alive Events, r=mccr8
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D232945
# ==============================================================================

diff -r 18c7bd8144dd -r f1e7580708c9 dom/events/Event.cpp
--- a/dom/events/Event.cpp	Fri Jan 10 18:49:56 2025 +0000
+++ b/dom/events/Event.cpp	Fri Jan 10 18:58:51 2025 +0000
@@ -29,6 +29,7 @@
 #include "mozilla/ViewportUtils.h"
 #include "mozilla/dom/Document.h"
 #include "mozilla/dom/DocumentInlines.h"
+#include "mozilla/dom/FragmentOrElement.h"
 #include "mozilla/dom/ShadowRoot.h"
 #include "mozilla/dom/WorkerScope.h"
 #include "mozilla/ScrollContainerFrame.h"
@@ -219,6 +220,40 @@
   NS_IMPL_CYCLE_COLLECTION_TRAVERSE(mOwner)
 NS_IMPL_CYCLE_COLLECTION_TRAVERSE_END
 
+NS_IMPL_CYCLE_COLLECTION_CAN_SKIP_BEGIN(Event)
+  if (tmp->HasKnownLiveWrapper()) {
+    if (WidgetEvent* event = tmp->mEvent) {
+      auto mark = [](EventTarget* aTarget) {
+        if (!aTarget) {
+          return;
+        }
+        if (nsINode* node = aTarget->GetAsNode()) {
+          FragmentOrElement::MarkNodeChildren(node);
+          if (node->HasKnownLiveWrapper()) {
+            // Use CanSkip to possibly mark more nodes to be certainly alive.
+            FragmentOrElement::CanSkip(node, true);
+          }
+        }
+      };
+
+      mark(event->mTarget);
+      mark(event->mCurrentTarget);
+      mark(event->mOriginalTarget);
+      mark(event->mRelatedTarget);
+      mark(event->mOriginalRelatedTarget);
+    }
+    return true;
+  }
+NS_IMPL_CYCLE_COLLECTION_CAN_SKIP_END
+
+NS_IMPL_CYCLE_COLLECTION_CAN_SKIP_IN_CC_BEGIN(Event)
+  return tmp->HasKnownLiveWrapperAndDoesNotNeedTracing(tmp);
+NS_IMPL_CYCLE_COLLECTION_CAN_SKIP_IN_CC_END
+
+NS_IMPL_CYCLE_COLLECTION_CAN_SKIP_THIS_BEGIN(Event)
+  return tmp->HasKnownLiveWrapper();
+NS_IMPL_CYCLE_COLLECTION_CAN_SKIP_THIS_END
+
 void Event::LastRelease() {
   nsISupports* supports = nullptr;
   QueryInterface(NS_GET_IID(nsCycleCollectionISupports),